### ISSUE TITLE
Installed check operators do not check against the same input to satisfy multiple conditions

### ISSUE THREAD

**Issue by icnocop** [dotnetinstaller/dotnetinstaller#163]
Installed check operators do not check against the same input to satisfy multiple conditions.

For example, give the following code snippet:
```
<installedcheckoperator type="And" description="Installed Check Operator">
    <installedcheck
        path="%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
        type="check_directory"
        comparison="version_ge"
        version="6.0.0"
        description="Install check if ASP.NET Core version 6.0.0 or later IS installed" />
    <installedcheck
        path="%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
        type="check_directory"
        comparison="version_lt"
        version="7.0.0"
        description="Install check if ASP.NET Core less than version 7.0.0 IS installed" />
</installedcheckoperator>
```

The installed check operator returns true if ASP.NET Core v3.1 and ASP.NET Core v8.0 are installed, even if ASP.NET Core v6.x is not installed.

I expected the installed check operator to return false, since ASP.NET Core v6.x isn't installed.

dni checks each `installedcheck` condition individually/separately.

In other words, the first condition will be evaluated as `true` because it will compare against `8.0.0`, and the second condition will be evaluated as `true` because it will compare against `3.1.32`.
`8.0.0` (the input) is not checked in both `installedcheck` conditions to determine the result of the parent `installedcheckoperator`.
Likewise, `3.1.32` (the input) is not checked in both `installedcheck` conditions to determine the result of the parent `installedcheckoperator`.

**Comment by dblock** [comment]
I am confused where 8.0.0 and 3.1.32 comes from. What is in`%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App`? Copy paste the log from these checks?

**Comment by icnocop** [comment]
Here are the contents of `%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App`:

![image](https://github.com/dotnetinstaller/dotnetinstaller/assets/507649/a350db79-9efc-46d4-8497-3ad4de856702)

In this case, the existence of these folders indicate that the ASP.NET Core 3.1.32 and ASP.NET Core 8.0.0 runtimes are installed.

Reference: https://learn.microsoft.com/en-us/dotnet/core/install/how-to-detect-installed-versions?pivots=os-windows#check-runtime-versions

I created a dotnetInstaller built pre-requisite installer which is configured to automatically download and install ASP.NET Core 6.0.25 if any ASP.NET Core 6.x runtime isn't detected as already being installed.
However, when ASP.NET Core 6 is not installed, the pre-requisite installer thinks it is.

%TEMP%\dotNetInstallerLog.txt:
```
2023-12-12 13:41:53	-------------------------------------------------------------------
2023-12-12 13:41:53	dotNetInstaller (DNI) started, version 3.2.2496.0
2023-12-12 13:41:53	Copyright (c) DevAge, Vestris Inc. & Contributors
2023-12-12 13:41:53	Operating system: Windows Server 2022 (x64)
2023-12-12 13:41:53	-------------------------------------------------------------------
2023-12-12 13:41:53	Sequence: install
2023-12-12 13:41:53	Loaded configuration from embedded resource 'RES_CONFIGURATION'
2023-12-12 13:41:53	Loaded schema: version=3.2.2496.0, generator=dotNetInstaller InstallerEditor
<snip>
2023-12-12 13:41:53	Loaded 'download' dialog component 'Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle', source=http://download.visualstudio.microsoft.com/download/pr/1fd87564-6bdb-4123-90dd-26488ec868c9/6c68988c310805bdcbb07b704fbe3e9d/dotnet-hosting-6.0.25-win.exe
2023-12-12 13:41:53	Loaded 'download' configuration 'Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle - Download Components' with 1 file(s)
2023-12-12 13:41:53	Loaded 'directory' installed check 'C:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App'
2023-12-12 13:41:53	Loaded 'directory' installed check 'C:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App'
2023-12-12 13:41:53	Loaded 'And' installed check operator
2023-12-12 13:41:53	Loaded Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle, display_name='Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle'
<snip>
2023-12-12 13:41:53	-- Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle (display_name='Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle', os_filter_lcid=, os_filter=, os_filter_min=, os_filter_max=, os_type_filter=, processor_architecture_filter=): LOADED
<snip>
2023-12-12 13:41:53	Check value '6.0.25' version no greater or equal match to '3.1.32'
2023-12-12 13:41:53	Check value '6.0.25' version greater or equal match to '8.0.0'
2023-12-12 13:41:53	Check value '7.0.0' version lesser match with '3.1.32'
2023-12-12 13:41:53	-- Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle (Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle): INSTALLED
<snip>
```

The dotNetInstaller.xml configuration looks similar to this (removed irrelevant attributes):
```
    <component id="Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle"
                    display_name="Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle">
        <downloaddialog>
            <download componentname="Microsoft ASP.NET Core 6.0.25 - Windows Hosting Bundle" />
        </downloaddialog>
        <installedcheckoperator type="And" description="Installed Check Operator">
            <installedcheck
                path="%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
                type="check_directory"
                comparison="version_ge"
                version="6.0.25"
                description="Install check if ASP.NET Core - Windows Hosting Bundle version 6.0.0 or later IS installed" />
            <installedcheck
                path="%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
                type="check_directory"
                comparison="version_lt"
                version="7.0.0"
                description="Install check if ASP.NET Core - Windows Hosting Bundle less than version 7.0.0 IS installed" />
        </installedcheckoperator>
    </component>
```

**Comment by dblock** [comment]
Is it missing `Check value '7.0.0' version lesser match with '8.0.0' above? I would check first whether this is by design or some side effect.

Next, it seems to me that you're looking for the inverse comparison, i.e. (3.1.32 >= 6.0.25 && 8.0.0 >= 6.0.25) (false) and (3.1.32 <= 7.0.0 && 8.0.0 <= 7.0.0) (false). Either the check is implemented incorrectly and we need to flip these, or it's implemented correctly and we need another version of it?

**Comment by icnocop** [comment]
> Is it missing `Check value '7.0.0' version lesser match with '8.0.0' above?

I don't think having this check matters, because it doesn't seem like there would be a way to use it which would change the expected results.

> I would check first whether this is by design or some side effect.

I think this is by design, and so I had previously added the `feature request` label to this issue.

> Next, it seems to me that you're looking for the inverse comparison, i.e. (3.1.32 >= 6.0.25 && 8.0.0 >= 6.0.25) (false) and (3.1.32 <= 7.0.0 && 8.0.0 <= 7.0.0) (false). Either the check is implemented incorrectly and we need to flip these, or it's implemented correctly and we need another version of it?

The issue is that the installed checks aren't evaluated like the `&&` operators as you indicated.

The left side and right side of the operators are evaluated independently.

For example, given the following configuration:
```
        <installedcheckoperator type="And" description="Installed Check Operator">
            <installedcheck
                path="%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
                type="check_directory"
                comparison="version_ge"
                version="6.0.25"
                description="Install check if ASP.NET Core - Windows Hosting Bundle version 6.0.0 or later IS installed" />
            <installedcheck
                path="%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
                type="check_directory"
                comparison="version_lt"
                version="7.0.0"
                description="Install check if ASP.NET Core - Windows Hosting Bundle less than version 7.0.0 IS installed" />
        </installedcheckoperator>
```
and only the following folders exist in `%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App`
```
3.1.32
8.0.0
```
then the configuration is evaluated like this
```
// first installed check:
// enumerate directories in "%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
//     found 3.1.32
//         compare 3.1.32 >= 6.0.25 (false)
//         LOG: Check value '6.0.25' version no greater or equal match to '3.1.32'
//     find next directory
//     found 8.0.0
//         compare 8.0.0 >= 6.0.25 (true)
//         LOG: Check value '6.0.25' version greater or equal match to '8.0.0'
//         return true

// second installed check:
// enumerate directories in "%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
//     found 3.1.32
//         compare 3.1.32 < 7.0.0 (true)
//         LOG: Check value '7.0.0' version lesser match with '3.1.32'
//         return true

// installed check operator
// return true
```

As you can see, the directories are enumerated twice, independently for each installed check.

**Comment by dblock** [comment]
Ok, that's correct, there's a version greater than 6.0 (8.0) and there's a version less than 7 (3.1). But we want one version to be greater than 6, and less than 7.

This shows a lot of limitations in the way these checks were designed. I think the cleanest and most generic way to solve this is to add a `semver` comparison (maybe `semver_all` and `semver_any` to allow for multiple combinations?). Thus, we could collapse these checks into one.

```
<installedcheck
                path="%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
                type="check_directory"
                comparison="semver_and"
                version=">= 6.0.0, < 7"
                description="Install check if ASP.NET Core - Windows Hosting Bundle version 6.x" />
```

**Comment by icnocop** [comment]
There's semantic version range syntax like `^6.0.0` used in popular package managers to permit versions from `6.0.0` all the way up to, but not including, the next major version, `7.0.0`.

However, the C++ semver parser and comparison code being used doesn't currently support this syntax.

A possible replacement (or addition) could be adapted from the code in https://github.com/easz/cpp-semver which seems to support advanced range syntax using the caret and tilde symbols.

I'm thinking this is a better approach than adding support for `semver_and`, and the additional parsing of comparison operators.

Instead, maybe we can add support for a comparison type of `semver_range`.

For example:
```
<installedcheck
                path="%PROGRAMFILES%\dotnet\shared\Microsoft.AspNetCore.App"
                type="check_directory"
                comparison="semver_range"
                version="^6.0.0"
                description="Install check if ASP.NET Core - Windows Hosting Bundle version 6.x is installed" />
```

What do you think?

**Comment by dblock** [comment]
Ah yes, 100%, I forgot about `^`, we should support all of the operators like `~`, `!`, etc.
